accessible/toolbox_treeview.component.js

```diff
@@ -0,0 +1,189 @@
+        <label #label id='{{treeService.createId(label)}}'>{{inputType(inputBlock.connection)}} {{valueOrStatement(inputBlock)}} needed:</label>
+        {{setLabelledBy(listItem1, concatStringWithSpaces('blockly-argument-text', label.id))}}
+      </li>
+    </div>
+  </ol>
+</li>
+<toolbox-tree-view *ngIf= 'block.nextConnection && block.nextConnection.targetBlock()' [level]='level' [block]='block.nextConnection.targetBlock()' [displayBlockMenu]='false'></toolbox-tree-view>
+    `,
+    directives: [ng.core.forwardRef(
+        function() { return blocklyApp.ToolboxTreeView; }), blocklyApp.FieldView],
+    inputs: ['block', 'displayBlockMenu', 'level', 'index', 'tree'],
+  })
+  .Class({
+    constructor: [blocklyApp.ClipboardService, blocklyApp.TreeService, function(_service, _service2) {
+      this.infoBlocks = {};
+      this.nextBlock = {};
+      this.sharedClipboardService = _service;
+      this.treeService = _service2;
+    }],
+    setLabelledBy: function(item,string){
+      if (!item.getAttribute('aria-labelledby')) {
+        item.setAttribute('aria-labelledby', string);
+      }
+    },
+    concatStringWithSpaces: function(){
+      var string = arguments[0];
+      for (i = 1; i < arguments.length; i++) {
+        var arg = arguments[i];
+        //arg can be undefined if it is coming from a function that checks whether the button should be disabled or not.
+        if (arg){
+          string = string + ' ' + arguments[i];
+        }
+      }
+      return string;
+    },
+    createCategoryDependantId: function(index, parentList){
+      //if this is the first block in a category-less toolbox, the id should be toolbox-tree-node0
+      if (index != undefined && index == 0) {
+        if (this.tree.getAttribute('aria-activedescendant') == 'blockly-toolbox-tree-node0') {
+          this.addClass(parentList, 'blocklyActiveDescendant');
+          parentList.setAttribute('aria-selected', 'true');
+          this.treeService.setActiveDesc(parentList, this.tree.id);
+        }
+        return 'blockly-toolbox-tree-node0';
+      } else {
+        return this.treeService.createId(parentList);
+      }
+    },
+    addClass: function(node, classText) {
+      //ensure that node doesn't have class already in it
+      var classList = node.className;
+      classList = classList.split(' ');
+      var canAdd = true;
+      for (var i=0; i<classList.length; i++){
+        if (classList[i] == classText) {
+          canAdd = false;
+        }
+      }
+      //add class if it doesn't
+      if (canAdd) {
+        if (classList.length == 0) {
+          node.className += classText;
+        } else {
+          node.className += (' ' + classText);
+        }
+      }
+    },
+    getInfo: function(block) {
+      //list all inputs
+      if (this.infoBlocks[block.id]) {
+        //this is required for some reason
```

Comment:
Um -- this comment raises alarm bells. What's the reason?

Is it because you need to keep references to the existing array? If so, explain why.